% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/antigen.garnish_predict.R
\name{garnish_affinity}
\alias{garnish_affinity}
\title{Perform neoepitope prediction.}
\usage{
garnish_affinity(dt = NULL, path = NULL, counts = NULL, min_counts = 1,
  assemble = TRUE, generate = TRUE, predict = TRUE, blast = fitness,
  fitness = TRUE, save = TRUE, remove_wt = TRUE)
}
\arguments{
\item{dt}{Data table. Input data table from \code{garnish_variants} or \code{garnish_jaffa}, or a data table in one of these forms:

dt with transcript id:\preformatted{Column name                 Example input

sample_id                   sample_1
ensembl_transcript_id       ENST00000311936
cDNA_change                 c.718T>A
MHC                         HLA-A*02:01 HLA-A*03:01
                            H-2-Kb H-2-Kb
                            HLA-DRB111:07 [second type]
}

dt with peptide (standard amino-acid one-letter codes only):\preformatted{Column name                 Example input

sample_id                   <same as above>
pep_mut                     MTEYKLVVVDAGGVGKSALTIQLIQNHFV
mutant_index                all
                            7
                            7 13 14
MHC                         <same as above>
}}

\item{path}{Path to input table (\href{https://cran.r-project.org/web/packages/rio/vignettes/rio.html#supported_file_formats}{acceptable formats}).}

\item{counts}{Optional. A file path to an RNA count matrix. The first column must contain ENSEMBL transcript ids. All samples in the input table must be present in the count matrix.}

\item{min_counts}{Integer. The minimum number of read counts that the transcript must be present to pass a variant. Default is 1.}

\item{assemble}{Logical. Assemble data table?}

\item{generate}{Logical. Generate peptides?}

\item{predict}{Logical. Predict binding affinities?}

\item{blast}{Logical. Run BLASTp to find wild-type peptide and known IEDB matches?}

\item{fitness}{Logical. Run model of \href{https://www.ncbi.nlm.nih.gov/pubmed/29132144}{Luksza et al. Nature 2017} to predict neoepitope fitness?}

\item{save}{Logical. Save a copy of garnish_affinity output to the working directory as "ag_output.txt"? Default is \code{TRUE}.}

\item{remove_wt}{Logical. Check all \code{nmer}s generated against wt peptidome and remove matches? Default is \code{TRUE}. If investigating wild-type sequences, set this to \code{FALSE}.}
}
\value{
A data table of binding predictions including:
\itemize{
\item \strong{cDNA_seq}: mutant cDNA sequence
\item \strong{cDNA_locs}: starting index of mutant cDNA
\item \strong{cDNA_locl}: ending index of mutant cDNA
\item \strong{cDNA_type}: netMHC prediction tool output
\item \strong{frameshift}: frameshift variant?
\item \strong{coding}: wt cDNA sequence
\item \strong{coding_mut}: mutant cDNA sequence
\item \strong{pep_type}: type of peptide
\item \strong{pep_mut}: mutant peptide sequence
\item \strong{pep_wt}: wt peptide sequence
\item \strong{mismatch_s}: starting index of mutant peptide sequence
\item \strong{mismatch_l}: ending index of mutant peptide sequence
\item \strong{mutant_index}: index of mutant peptide
\item \strong{nmer}: nmer for prediction
\item \strong{nmer_i}: index of nmer in sliding window
\item \strong{*_net}: netMHC prediction tool output
\item \strong{mhcflurry_*}: mhcflurry_ prediction tool output
\item \strong{mhcnuggets_*}: mhcnuggets_ prediction tool output
\item \strong{DAI}: Differential agretopicty index of missense and corresponding wild-type peptide, see \code{garnish_summary} for an explanation of DAI.
\item \strong{BLAST_A}: Ratio of consensus binding affinity of mutant peptide / closest single AA mismatch from blastp results. Returned only if \code{blast = TRUE}.
}

clonality info:
\itemize{
\item \strong{clone_id}: rank of the clone containing the variant (highest equals larger tumor fraction).
\item \strong{cl_proportion}: The estimated mean tumor fraction containing the clone. If allele fraction and not clonality is used, this is estimated.
}

antigen.garnish fitness model results
\itemize{
\item \strong{Consensus_scores}: average value of MHC binding affinity from all prediction tools that contributed output. 95\% confidence intervals are given by \strong{Upper_CI}, \strong{Lower_CI}.
\item \strong{iedb_score}: R implementation of TCR recognition probability for peptide through summing of alignments in IEDB for corresponding organism.
\item \strong{min_DAI}: Minimum of value of BLAST_A or DAI values, to provide the most conservative estimate differential binding between input and wildtype matches.
\item \strong{fitness_score}: Product of min_DAI and iedb_score. The peptide with the highest value per clone is the top neoepitope. Does not apply to wildtype input.
\item \strong{garnish_score}: overall sample immunogenicity based on top clonal neoepitopes. Only if clonality or allele fraction data is present in the table. See ?garnish_variants.
}

fitness model information \href{https://www.ncbi.nlm.nih.gov/pubmed/29132144}{Luksza et al. Nature 2017}:
\itemize{
\item \strong{NeoantigenRecognitionPotential}: Product of A and R (amplitude, analogous to min_DAI, and TCR recognition probability components).
\item full raw output saved to the working directory.
}

transcript description:
\itemize{
\item description
\item start_position
\item end_position
\item transcript_end
\item transcript_length
\item transcript_start
\item peptide
}
}
\description{
Perform ensemble neoepitope prediction on a data table of missense mutations, insertions, deletions or gene fusions using netMHC, mhcflurry, and mhcnuggets.
}
\details{
\itemize{
\item see \code{list_mhc} for compatible MHC allele syntax
multiple MHC alleles for a single sample_id should be space separated. Murine and human alleles should be in separate rows
\item \code{garnish_score} is calculated if allelic fraction or tumor cellular fraction were provided
}
}
\examples{
\dontrun{
library(magrittr)
library(antigen.garnish)

# run full pipeline

   # download an example VCF
   dt <- "antigen.garnish_example.vcf" \%T>\%
   utils::download.file("http://get.rech.io/antigen.garnish_example.vcf", .) \%>\%

   # extract variants
   antigen.garnish::garnish_variants \%>\%

   # add space separataed MHC types with antigen.garnish::list_mhc() compatible names
       .[, MHC := c("HLA-A*02:01 HLA-DRB1*14:67",
                   "H-2-Kb H-2-IAd",
                   "HLA-A*01:47 HLA-DRB1*03:08")] \%>\%

   # predict neoepitopes
   antigen.garnish::garnish_affinity \%>\%

   # summarize predictions
   antigen.garnish::garnish_summary \%T>\%
   print

}

\dontrun{
# input a data table of transcripts

library(magrittr)
library(antigen.garnish)

 dt <- data.table::data.table(
          sample_id = "test",
          ensembl_transcript_id =
          c("ENSMUST00000128119",
            "ENSMUST00000044250",
            "ENSMUST00000018743"),
          cDNA_change = c("c.4988C>T",
                          "c.1114T>G",
                          "c.718T>A"),
          MHC = c("HLA-A*02:01 HLA-DRB1*14:67",
                  "H-2-Kb H-2-IAd",
                  "HLA-A*01:47 HLA-DRB1*03:08")) \%>\%
 antigen.garnish::garnish_affinity \%T>\%
 str
}

\dontrun{
# input a data table of peptides for all MHC types

library(data.table)
library(magrittr)
library(antigen.garnish)

 dt <- data.table::data.table(
          sample_id = "test",
          pep_mut = "MTEYKLVVVGAGDVGKSALTIQLIQNHFVDEYDP",
          mutant_index = "12",
          MHC = "all") \%>\%
 antigen.garnish::garnish_affinity \%T>\%
 str
}

\dontrun{
# input from Microsoft excel

   # download an example excel file
     path <- "antigen.garnish_test_input.xlsx" \%T>\%
       utils::download.file("http://get.rech.io/antigen.garnish_test_input.xlsx", .)

   # predict neoepitopes
     dt <- antigen.garnish::garnish_affinity(path = path) \%T>\%
     str
}

}
\references{
Luksza, M, Riaz, N, Makarov, V, Balachandran VP, et al. 2017. A neoepitope fitness model predicts tumour response to checkpoint blockade immunotherapy. Nature. 23;551(7681):512-516
}
\seealso{
\code{\link{garnish_variants}}

\code{\link{list_mhc}}

\code{\link{garnish_summary}}
}
