% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/antigen.garnish_predict.R
\name{garnish_predictions}
\alias{garnish_predictions}
\title{Performs epitope prediction.}
\usage{
garnish_predictions(dt, assemble = TRUE, generate = TRUE, predict = TRUE,
  humandb = "GRCh38", mousedb = "GRCm38")
}
\arguments{
\item{dt}{Data table. Input data table from garnish_variants or a data table in one of these forms:

dt with transcript id:\preformatted{Column name                 Example input

sample_id                   sample_1
ensembl_transcript_id       ENST00000311936
cDNA_change                 c.718T>A
MHC                         HLA-A02:01 HLA-A03:01
                            H-2-Kb H-2-Kb
                            HLA-DRB111:07 [second type]
}

dt with peptide:\preformatted{Column name                 Example input

sample_id                   <same as above>
pep_mut                     MTEYKLVVVDAGGVGKSALTIQLIQNHFV
mutant_index                all
                            7
                            7 13 14
MHC                         <same as above>
}}

\item{assemble}{Logical. Assemble data table?}

\item{generate}{Logical. Generate peptides?}

\item{predict}{Logical. Predict binding affinities?}

\item{humandb}{Character vector. One of "GRCh37" or "GRCh38".}

\item{mousedb}{Character vector. One of "GRCm37" or "GRCm38".}
}
\value{
A data table of binding predictions including:
\itemize{
\item \strong{cDNA_seq}: mutant cDNA sequence
\item \strong{cDNA_locs}: starting index of mutant cDNA
\item \strong{cDNA_locl}: ending index of mutant cDNA
\item \strong{cDNA_type}: netMHC prediction tool output
\item \strong{frameshift}: frameshift variant?
\item \strong{coding}: wt cDNA sequence
\item \strong{coding_mut}: mutant cDNA sequence
\item \strong{pep_type}: type of peptide for this row
\item \strong{pep_mut}: mutant cDNA sequence
\item \strong{pep_wt}: wt cDNA sequence
\item \strong{mismatch_s}: starting index of mutant peptide sequence
\item \strong{mismatch_l}: ending index of mutant peptide sequence
\item \strong{mutant_index}: index of mutant peptide for this row
\item \strong{nmer}: nmer for prediction
\item \strong{nmer_i}: index of nmer in sliding window
\item \strong{*_net}: netMHC prediction tool output
\item \strong{mhcflurry_*}: netMHC prediction tool output
}

as well as a transcript description:
\itemize{
\item description
\item start_position
\item end_position
\item transcript_end
\item transcript_length
\item transcript_start
\item peptide
\item refseq_mrna
}
}
\description{
Performs epitope prediction on a data table of missense mutations.
}
\examples{
\dontrun{
library(magrittr)
library(antigen.garnish)

# full pipeline

  # download an example VCF
    dt <- "antigen.garnish_example.vcf" \%T>\%
    utils::download.file("http://get.rech.io/antigen.garnish_example.vcf", .) \%>\%

  # extract variants
    garnish_variants \%>\%

  # predict neoepitopes
    garnish_predictions \%T>\%
    str

# peptide input

  # create input data table
    dt <- data.table::data.table(
           sample_id = "test",
           pep_mut = "ATGACTGAATATAAACTTGTGGTA",
           mutant_index = "7 13 14",
           MHC = "H-2-Kb"

    dto <- garnish_predictions(dt)

}

\dontrun{
# using an existing data table

library(data.table)
library(magrittr)
library(antigen.garnish)

 dt <- data.table::data.table(
          sample_id = "test",
          ensembl_transcript_id =
          c("ENSMUST00000128119",
            "ENSMUST00000044250",
            "ENSMUST00000018743"),
          cDNA_change = c("c.4988C>T",
                          "c.1114T>G",
                          "c.718T>A"),
          MHC = c("HLA-A*02:01 HLA-DRB1*14:67",
                  "H-2-Kb H-2-IAd",
                  "HLA-A*01:47 HLA-DRB1*03:08")) \%>\%
 garnish_predictions \%T>\%
 str
}
}
